### PR TITLE
Information Retrieval (3/5)

### PR DESCRIPTION
**Pull request by lucadiliello**
## Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the contributor guideline, Pull Request section?
- [X] Did you make sure to update the docs?
- [X] Did you write any new necessary tests?

## What does this PR do?

Implements Precision (@K) for Information Retrieval. More info [here](https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Precision).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by lucadiliello** [comment]
Changed also the test of MRR to compare against the [sklearn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.label_ranking_average_precision_score.html?highlight=precision#sklearn.metrics.label_ranking_average_precision_score) instead of mine.
For Precision, I didn't find a reliable implementation :(

File: tests/functional/test_retrieval.py [hunk 1]
```diff
@@ -93,3 +126,19 @@ def test_input_shapes(torch_metric) -> None:
 
     with pytest.raises(ValueError, match="`preds` and `target` must be of the same shape"):
         torch_metric(preds, target)
+
+
+# test metrics using top K parameter
+@pytest.mark.parametrize(['torch_metric'], [
+    [retrieval_precision]
+])
```
Comment (SkafteNicki) [inline tests/functional/test_retrieval.py:134]: no need to use parametrization if only single input
```suggestion

])
```

File: tests/retrieval/helpers.py [hunk 2]
```diff
@@ -34,10 +34,15 @@ def _compute_sklearn_metric(
 
 
 def _test_retrieval_against_sklearn(
-    sklearn_metric, torch_metric, size, n_documents, query_without_relevant_docs_options
+    sklearn_metric,
+    torch_metric,
+    size,
+    n_documents,
+    query_without_relevant_docs_options,
```
Comment (SkafteNicki) [inline tests/retrieval/helpers.py:41]: not sure if they are correct
```suggestion
    sklearn_metric: Callable,
    torch_metric: Metric,
    size: int,
    n_documents: int,
    query_without_relevant_docs_options: str,
```

File: tests/retrieval/helpers.py [hunk 3]
```diff
@@ -120,3 +125,9 @@ def _test_input_shapes(torchmetric) -> None:
 
     with pytest.raises(ValueError, match="`indexes`, `preds` and `target` must be of the same shape"):
         metric(indexes, preds, target)
+
+
+def _test_input_args(torchmetric, message, **kwargs) -> None:
```
Comment (SkafteNicki) [inline tests/retrieval/helpers.py:130]:
```suggestion
def _test_input_args(torchmetric: Metric, message: str, **kwargs) -> None:
```

**Review by SkafteNicki — APPROVED** [review]
smaller comments else LGTM

File: tests/functional/test_retrieval.py (hunk 1 above)
Comment (lucadiliello) [reply tests/functional/test_retrieval.py:134]: It's future proof. In the next PR I will add recall:
```
@pytest.mark.parametrize(['torch_metric'], [
        [retrieval_precision],
        [retrieval_recall],
    ])
```

**Comment by SkafteNicki** [comment]
@lucadiliello tests seems to be failing (missing import I think). Could you look into it? :]

**Review by Borda — APPROVED** [review]

File: torchmetrics/retrieval/mean_reciprocal_rank.py [hunk 4]
```diff
@@ -21,7 +20,7 @@
 class RetrievalMRR(RetrievalMetric):
     """
     Computes `Mean Reciprocal Rank
-    <https://en.wikipedia.org/wiki/Mean_reciprocal_rank>`_.
+    <https://en.wikipedia.org/wiki/Mean_reciprocal_rank>`__.
```
[the remark this answers is not in the document]
Comment (lucadiliello) [reply torchmetrics/retrieval/mean_reciprocal_rank.py:23]: I receive an error when doing `make docs` if I use a single underscore. This is probably caused by having the same link to wikipedia 2 times: 1 link in the Metric class and one in the functional version.
The error is:

>     /Users/lucadiliello/development/metrics/torchmetrics/functional/retrieval/reciprocal_rank.py:docstring of torchmetrics.functional.retrieval.reciprocal_rank.retrieval_reciprocal_rank:22: WARNING: Duplicate explicit target name: "here".

So I found that by using 2 underscores it [does work](https://thomas-cokelaer.info/blog/2020/04/sphinx-raise-a-warning-duplicate-explicit-target-name/).
